aiida/manage/tests/pytest_fixtures.py

```diff
@@ -0,0 +1,145 @@
+
+
+@pytest.fixture(scope='function')
+def clear_database(aiida_profile):  # pylint: disable=redefined-outer-name
+    """Clear the database after each test.
+
+    Note: Thanks to ``autouse=True``, this happens as soon as you import this fixure in your ``conftest.py``.
+    """
+    yield
+    # after the test function has completed, reset the database
+    aiida_profile.reset_db()
+
+
+@pytest.fixture(scope='function')
+def temp_dir():
+    """Get a temporary directory.
+
+    E.g. to use as the working directory of an AiiDA computer.
+
+    :return: The path to the directory
+    :rtype: str
+
+    """
+    dirpath = tempfile.mkdtemp()
+    yield dirpath
+    # after the test function has completed, remove the directory again
+    shutil.rmtree(dirpath)
+
+
+@pytest.fixture(scope='function')
+def aiida_localhost(temp_dir):  # pylint: disable=redefined-outer-name
+    """Get an AiiDA computer for localhost.
+
+    Usage::
+
+      def test_1(aiida_localhost):
+          name = aiida_localhost.get_name()
+          # proceed to set up code or use 'aiida_local_code_factory' instead
+
+
+    :return: The computer node
+    :rtype: :py:class:`aiida.orm.Computer`
+    """
+    from aiida.orm import Computer
+    from aiida.common.exceptions import NotExistent
+
+    name = 'localhost'
+
+    try:
+        computer = Computer.objects.get(name=name)
+    except NotExistent:
+        computer = Computer(
+            name=name,
+            description='localhost computer set up by test manager',
+            hostname=name,
+            workdir=temp_dir,
+            transport_type='local',
+            scheduler_type='direct'
+        )
+        computer.store()
+        computer.configure()
+
+    return computer
+
+
+@pytest.fixture(scope='function')
+def aiida_local_code_factory(aiida_localhost):  # pylint: disable=redefined-outer-name
+    """Get an AiiDA code on localhost.
+
+    Searches in the PATH for a given executable and creates an AiiDA code with provided entry point.
+
+    Usage::
+
+      def test_1(aiida_local_code_factory):
+          code = aiida_local_code_factory('pw.x', 'quantumespresso.pw')
+          # use code for testing ...
+
+    :return: A function get_code(executable, entry_point) that returns the Code node.
+    :rtype: object
+    """
+
+    def get_code(executable, entry_point, computer=aiida_localhost):
+        """Get local code.
+        Sets up code for given entry point on given computer.
+
+        :param entry_point: Entry point of calculation plugin
+        :param computer: (local) AiiDA computer
+        :return: The code node
+        :rtype: :py:class:`aiida.orm.Code`
+        """
+        from aiida.orm import Code
+        from aiida.common.files import which
+
+        codes = Code.objects.find(filters={'label': executable})  # pylint: disable=no-member
+        if codes:
+            return codes[0]
+
+        executable_path = which(executable)
```

Comment:
What happens if `which` cannot find it? I also use a similar fixture to create a `Code` instance on localhost, but I actually never intend to run it. I just need the `Code` object. This fixture would make that impossible. For reference this is what I use:
```
@pytest.fixture
def generate_code_localhost():
    """Return a `Code` instance configured to run calculations of given entry point on localhost `Computer`."""

    def _generate_code_localhost(entry_point_name, computer):
        from aiida.orm import Code
        plugin_name = entry_point_name
        remote_computer_exec = [computer, '/bin/true']
        return Code(input_plugin_name=plugin_name, remote_computer_exec=remote_computer_exec)

    return _generate_code_localhost
```
Can we make it optional of the `executable` really having to correspond to a physical executable on the machine?

Reply:
I think if an executable is specified and not found, the fixture should fail.

One could think of setting the default value of the `executable` parameter to `None` and then doing something like
```python
        if executable is None:
            executable_path = '/fake/path' 
        else:        
            executable_path = which(executable)
```

I would be ok with this - still, it feels a bit like cheating since we are taking advantage of the fact that the `Code` constructor does not check whether the path that is provided actually exists.
There is actually an issue https://github.com/aiidateam/aiida-core/issues/868 (opened by you ;-) ) to fix this - after that, this solution would no longer work.

Let me know what you think.

Reply:
I see, although in reference to #868 I now don't think that in the case of a remote machine, code creation should validate the path, given that this would have to open transport adding a lot more complexity and potential for errors. For local codes and codes on localhost computers there is still something to be said for this check and this would amount to your `which` check in the tests. 

That is why I was using `/bin/true` though, which would at least work on typical Linux and MacOS systems. I still think that there should be a way of creating a `Code` instance for testing purposes that does not necessarily have an actual existing path. All the testing in `aiida-quantumespresso` is done this way, so I think this is really valuable and it is not necessary to actually run the calculations to provide useful testing. Would your current implementation work if I pass `/bin/true`, I think so right?

Reply:
> Would your current implementation work if I pass /bin/true, I think so right?

Yes - by the way, `/bin/true` does not exist on my system:

```python
In [1]: from aiida.common.files import which

In [2]: which('true')
Out[2]: '/opt/local/libexec/gnubin/true'

In [3]: which('/bin/true')

In [4]: which('/opt/local/libexec/gnubin/true')
Out[4]: '/opt/local/libexec/gnubin/true'
```